File: Leetcode/0054 Spiral Matrix.java
class Solution {
    private static final int[][] DIRS = {
            {0, 1},
            {1, 0},
            {0, -1},
            {-1, 0},
    };

    private static final int VISITED = 101;

    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> spiral = new ArrayList<>();
        int n = matrix.length, m = matrix[0].length, number = 1;
        int elements = n * m, dirIndex = 0;
        int i = 0, j = 0;
        while (number <= elements) {
            spiral.add(matrix[i][j]);
            matrix[i][j] = VISITED;
            int nextI = i + DIRS[dirIndex][0], nextJ = j + DIRS[dirIndex][1];
            boolean isInvalid = 
                nextI < 0 || nextI >= n 
                || nextJ < 0 || nextJ >= m 
                || matrix[nextI][nextJ] == VISITED;
            if (isInvalid) {
                dirIndex = (dirIndex + 1) % DIRS.length;
            }
            i += DIRS[dirIndex][0];
            j += DIRS[dirIndex][1];
            ++number;
        }
        return spiral;
    }
}

File: Leetcode/0054 Spiral Matrix.ts
const DIRS = [[0, 1], [1, 0], [0, -1], [-1, 0]];
const VISITED = -101;

function spiralOrder(matrix: number[][]): number[] {
    const n = matrix.length, m = matrix[0].length, elems = n * m;
    const spiral: number[] = [];
    let i = 0, j = 0, dirIndex = 0;
    while (spiral.length < elems) {
        spiral.push(matrix[i][j]);
        matrix[i][j] = VISITED;
        const nextI = i + DIRS[dirIndex][0], nextJ = j + DIRS[dirIndex][1];
        const isInvalid = nextI < 0 || nextI >= n || nextJ < 0 || nextJ >= m || matrix[nextI][nextJ] === VISITED;
        if (isInvalid) {
            dirIndex = (dirIndex + 1) % DIRS.length;
        }
        i = i + DIRS[dirIndex][0];
        j = j + DIRS[dirIndex][1];
    }
    return spiral;
}
